File: backend/api/organization.go
package api

import (
	"emviwiki/backend/context"
	"emviwiki/backend/organization"
	"emviwiki/shared/rest"
	"net/http"
)

const (
	maxOrganizationPictureMem = 2097152 // 2 MBi
)

func GetOrganizationsHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	orgas := organization.ReadOrganizations(ctx.UserId)

	for i := range orgas {
		if orgas[i].Picture.Valid {
			orgas[i].Picture.SetValid(getResourceURL(orgas[i].Picture.String))
		}
	}

	rest.WriteResponse(w, orgas)
	return nil
}

func GetOrganizationHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	if ctx.Organization != nil {
		if ctx.Organization.Picture.Valid {
			ctx.Organization.Picture.SetValid(getResourceURL(ctx.Organization.Picture.String))
		}

		rest.WriteResponse(w, ctx.Organization)
		return nil
	}

	orga, err := organization.ReadOrganization(ctx)

	if err != nil {
		return []error{err}
	}

	if orga.Picture.Valid {
		orga.Picture.SetValid(getResourceURL(orga.Picture.String))
	}

	rest.WriteResponse(w, orga)
	return nil
}

func CreateOrganizationHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	req := organization.CreateOrganizationData{}

	if err := rest.DecodeJSON(r, &req); err != nil {
		return []error{err}
	}

	if errs := organization.CreateOrganization(ctx.UserId, req); errs != nil {
		return errs
	}

	return nil
}

func ValidateOrganizationHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	req := struct {
		organization.CreateOrganizationData

		ValidateNameDomain   bool `json:"validate_name_domain"`
		ValidateUsernameLang bool `json:"validate_username_lang"`
	}{}

	if err := rest.DecodeJSON(r, &req); err != nil {
		return []error{err}
	}

	req.Trim()
	errs := make([]error, 0)

	if req.ValidateNameDomain {
		if err := req.CheckDomainValid(req.Domain); err != nil {
			errs = append(errs, err)
		}

		if err := req.CheckNameValid(req.Name); err != nil {
			errs = append(errs, err)
		}
	} else if req.ValidateUsernameLang {
		if err := req.CheckUsernameValid(req.Username); err != nil {
			errs = append(errs, err)
		}

		if err := req.CheckLanguage(); err != nil {
			errs = append(errs, err)
		}
	}

	if len(errs) != 0 {
		return errs
	}

	return nil
}

func DeleteOrganizationHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	name := r.URL.Query().Get("name")

	if err := organization.DeleteOrganization(ctx.Organization.ID, ctx.UserId, name, authProvider); err != nil {
		return []error{err}
	}

	return nil
}

func UploadOrganizationPictureHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	r.Body = http.MaxBytesReader(w, r.Body, maxOrganizationPictureMem)

	if err := organization.UploadOrganizationPicture(r, ctx.Organization.ID, ctx.UserId); err != nil {
		return []error{err}
	}

	return nil
}

func DeleteOrganizationPictureHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	if err := organization.DeleteOrganizationPicture(ctx.Organization.ID, ctx.UserId); err != nil {
		return []error{err}
	}

	return nil
}

func UpdateOrganizationHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	req := struct {
		Name   string `json:"name"`
		Domain string `json:"domain"`
	}{}

	if err := rest.DecodeJSON(r, &req); err != nil {
		return []error{err}
	}

	if err := organization.UpdateOrganization(ctx.Organization, ctx.UserId, req.Name, req.Domain); err != nil {
		return err
	}

	return nil
}

func UpdateOrganizationPermissionsHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	req := struct {
		CreateGroupAdmin bool `json:"create_group_admin"`
		CreateGroupMod   bool `json:"create_group_mod"`
	}{}

	if err := rest.DecodeJSON(r, &req); err != nil {
		return []error{err}
	}

	if err := organization.UpdateOrganizationPermissions(ctx.Organization, ctx.UserId, req.CreateGroupAdmin, req.CreateGroupMod); err != nil {
		return []error{err}
	}

	return nil
}

func GetOrganizationStatisticsHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	statistics, err := organization.GetOrganizationStatistics(ctx.Organization, ctx.UserId)

	if err != nil {
		return []error{err}
	}

	rest.WriteResponse(w, statistics)
	return nil
}

func GenerateInvitationCodeHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	req := struct {
		ReadOnly bool `json:"read_only"`
	}{}

	if err := rest.DecodeJSON(r, &req); err != nil {
		return []error{err}
	}

	code, err := organization.GenerateInvitationCode(ctx.Organization, ctx.UserId, req.ReadOnly)

	if err != nil {
		return []error{err}
	}

	rest.WriteResponse(w, struct {
		Code string `json:"code"`
	}{code})
	return nil
}

func GetInvitationCodeHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	code, readOnly, err := organization.GetInvitationCode(ctx.Organization, ctx.UserId)

	if err != nil {
		return []error{err}
	}

	rest.WriteResponse(w, struct {
		Code     string `json:"code"`
		ReadOnly bool   `json:"read_only"`
	}{code, readOnly})
	return nil
}

File: frontend/public/src/service/organization.js
import axios from "axios";

export const OrganizationService = new class {
	getOrganization() {
		return new Promise((resolve, reject) => {
			axios.get(`${EMVI_WIKI_BACKEND_HOST}/api/v1/organization`)
			.then((r) => {
				resolve(r.data);
			})
			.catch((e) => {
				reject(e);
			});
		});
	}

	leaveOrganization(name) {
		return new Promise((resolve, reject) => {
			axios.post(`${EMVI_WIKI_BACKEND_HOST}/api/v1/organization/exit`, {name})
			.then((r) => {
				resolve(r.data);
			})
			.catch((e) => {
				reject(e);
			});
		});
	}

	deleteOrganization(name) {
		return new Promise((resolve, reject) => {
			axios.delete(`${EMVI_WIKI_BACKEND_HOST}/api/v1/organization`, {params: {name}})
			.then((r) => {
				resolve(r.data);
			})
			.catch((e) => {
				reject(e);
			});
		});
	}

	uploadPicture(form) {
		return new Promise((resolve, reject) => {
			axios.post(`${EMVI_WIKI_BACKEND_HOST}/api/v1/organization/picture`, form)
			.then(r => {
				resolve(r);
			})
			.catch(e => {
				reject(e);
			})
		});
	}

	deletePicture() {
		return new Promise((resolve, reject) => {
			axios.delete(`${EMVI_WIKI_BACKEND_HOST}/api/v1/organization/picture`)
			.then(r => {
				resolve(r);
			})
			.catch(e => {
				reject(e);
			})
		});
	}

	updateOrganization(name, domain) {
		return new Promise((resolve, reject) => {
			axios.put(`${EMVI_WIKI_BACKEND_HOST}/api/v1/organization`, {name, domain})
			.then(r => {
				resolve(r);
			})
			.catch(e => {
				reject(e);
			})
		});
	}

	updateOrganizationPermissions(create_group_admin, create_group_mod) {
		return new Promise((resolve, reject) => {
			axios.put(`${EMVI_WIKI_BACKEND_HOST}/api/v1/organization/permissions`, {create_group_admin, create_group_mod})
			.then(r => {
				resolve(r);
			})
			.catch(e => {
				reject(e);
			})
		});
	}

	getStatistics() {
		return new Promise((resolve, reject) => {
			axios.get(`${EMVI_WIKI_BACKEND_HOST}/api/v1/organization/statistics`)
			.then(r => {
				resolve(r.data);
			})
			.catch(e => {
				reject(e);
			})
		});
	}

	generateInvitationCode(read_only) {
		return new Promise((resolve, reject) => {
			axios.post(`${EMVI_WIKI_BACKEND_HOST}/api/v1/organization/invitation`, {read_only})
				.then(r => {
					resolve(r.data.code);
				})
				.catch(e => {
					reject(e);
				})
		});
	}

	getInvitationCode() {
		return new Promise((resolve, reject) => {
			axios.get(`${EMVI_WIKI_BACKEND_HOST}/api/v1/organization/invitation`)
				.then(r => {
					resolve({code: r.data.code, read_only: r.data.read_only});
				})
				.catch(e => {
					reject(e);
				})
		});
	}
};
